### PR TITLE
Adding support for spectral subsets for model fitting in cubeviz

### PR DESCRIPTION
**Pull request by bmorris3**
### Background

@orifox pointed out a bug for Cubeviz users. When fitting a model to a spectral cube, the user can select a spectral subset from the "Spectral region" dropdown. Then when the user clicks Fit Model, Cubeviz fits the model to the full cube, not respecting the selected spectral subset. 

### Diagnosis

The problem arises because:

1. Cubeviz creates a `Spectrum1D` object for every spaxel without preserving the original mask:

https://github.com/spacetelescope/jdaviz/blob/10fbd3bfde3694faa0e984fa5c89cef7bc820986/jdaviz/configs/default/plugins/model_fitting/fitting_backend.py#L254

2. When fitting a spectrum with `ModelFitting._fit_model_to_spectrum` (_not_ the _cube_), Cubeviz knows to update the mask of the `Spectrum1D` object used internally:

https://github.com/spacetelescope/jdaviz/blob/10fbd3bfde3694faa0e984fa5c89cef7bc820986/jdaviz/configs/default/plugins/model_fitting/model_fitting.py#L663-L672

but there is no analogous `mask` block in `ModelFitting._fit_model_to_cube`. This might be because in the `_cube` version, the `window` kwarg is passed, which pushes the spectral region-handling responsibilities to `specutils.fitting.fit_lines` ([docs](https://specutils.readthedocs.io/en/stable/api/specutils.fitting.fit_lines.html)):

https://github.com/spacetelescope/jdaviz/blob/10fbd3bfde3694faa0e984fa5c89cef7bc820986/jdaviz/configs/default/plugins/model_fitting/model_fitting.py#L740-L745

If that was intended when `_cube` was written, it wasn't applied uniformly, because `ModelFitting` doesn't give `fit_lines` a `window`: 

https://github.com/spacetelescope/jdaviz/blob/10fbd3bfde3694faa0e984fa5c89cef7bc820986/jdaviz/configs/default/plugins/model_fitting/model_fitting.py#L675-L679

### Description

This PR restores the intended behavior, and makes Cubeviz more internally consistent when fitting a spectrum vs. a cube. Now when a spectral subset is selected, only that subset is included in the fit. I've added a `mask` in the `Spectrum1D` constructor, which is respected by the specutils fitting functions. 

I also:
* added the `window` kwarg where it was missing in `ModelFitting._fit_model_to_spectrum`
* added a test to check that the mask is preserved in the fitted spectrum
* cleaned up some typos

[🐱](https://jira.stsci.edu/browse/JDAT-2918)

*Update*: this PR now also implements the similarly implied-but-missing feature for spatial subset selection in Cubeviz fits.

### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)? INC0183815

### CODE REVIEW

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.vue
Comment (bmorris3) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.vue:199]: The change to this message wording isn't strictly necessary for this PR, but I found it helpful when learning to use this workflow.

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.py:758]: if this is copy-pasted, can we put it somewhere that both spectrum and cube-fitting can use?

Comment (bmorris3) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py:758]: It's close, but slightly different than the related passage here:

https://github.com/spacetelescope/jdaviz/blob/10fbd3bfde3694faa0e984fa5c89cef7bc820986/jdaviz/configs/default/plugins/model_fitting/model_fitting.py#L663-L672

I'll see if I can do that. 👍🏻

Comment (bmorris3) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py]: Done!

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.py:663]:
```suggestion
        # Apply mask from selected spectral subset
```

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.py:733]:
```suggestion
        # Apply mask from selected spectral subset
```

**Review by kecnry — COMMENTED** [review]
Code changes look nice and clean!  Just to make sure I'm understanding correctly, the implementation under-the-hood supports `window`, but we don't make use of that and use masks instead from the plugin?

Is it safe to say that spatial subsets are ignored for cubes?  Is that the behavior we want?  Do we need any UI changes to make that more clear?

**Comment by bmorris3** [comment]
> Just to make sure I'm understanding correctly, the implementation under-the-hood supports `window`, but we don't make use of that and use masks instead from the plugin?

Correct. I talked this over with @rosteen, who told me: 

```
    I'm suspicious that the window handling is actually only 
    taking the first region if it's given a list of SpectralRegions 
    and a single model 
```

So I've left the `window` keyword arguments here in case they provide some functionality that I haven't yet imagined, rather than deleting them.

> Is it safe to say that spatial subsets are ignored for cubes? 

I _think_ this is true. Maybe @rosteen knows the history a bit better?

> Is that the behavior we want? Do we need any UI changes to make that more clear?

We give the user a dropdown for selecting the full cube or a spatial region: 

https://github.com/spacetelescope/jdaviz/blob/85b8d2ca5346aa13e87ac441cd466425a7227618/jdaviz/configs/default/plugins/model_fitting/model_fitting.vue#L22

even though we don't seem to use that information. It might be as simple as adding an `if mask` condition in the list comprehension that lists the spaxels to fit here:

https://github.com/spacetelescope/jdaviz/blob/85b8d2ca5346aa13e87ac441cd466425a7227618/jdaviz/configs/default/plugins/model_fitting/fitting_backend.py#L353-L373

How does that sound?

**Comment by kecnry** [comment]
I think that sounds reasonable (we should either also support spatial subsets, if feasible even though technically not reported in this bug report, or a big warning or disable the fit button if cube fit is enabled and a spatial subset is selected in the dropdown).

File: CHANGES.rst
Comment (pllim) [inline CHANGES.rst:102]: This should be moved under Cubeviz section below.
```suggestion
- Support for fitting spectral subsets with Cubeviz. [#1834]
```

File: jdaviz/configs/default/plugins/model_fitting/fitting_backend.py
Comment (pllim) [inline jdaviz/configs/default/plugins/model_fitting/fitting_backend.py:264]: What if `mask = self.mask[x, y, :]` is all `True`? Is that possible? In that case, what is the point running this? Should this abort early in that case?

Or should we only extract the unmasked flux here instead of passing in `mask` or is `specutils` smart enough to optimize resources if a big array is passed in but most of it is masked out?

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (pllim) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.py:808]: Coverage is complaining that this is not tested.

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.py:731]: Does this support either the subset name or cls as the first positional argument?

```suggestion
            spec = data.get_object(cls=Spectrum1D, statistic=None)
```

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.py:830]: can we move this up higher to `SubsetSelect.selected_subset_data` (or something like that)?  Then you can rely on the component to do all the viewer accessing as well and hide all the ugly code somewhere that is a bit more reusable.

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.py:848]: any downside to adding (vs multiplying)?  I see this is how it was previously as well, but now that we're combining spatial and spectral, this code might actually get triggered whereas it probably wasn't before unless a mask already existed on the input data.

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (bmorris3) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py:731]: There's a helpful demo [here](https://glue-astronomy.readthedocs.io/en/latest/translators.html#datasets-and-subsets): `get_object` takes `cls`, but `get_subset_object` takes the subset as the first argument, then `cls`.

File: jdaviz/configs/default/plugins/model_fitting/fitting_backend.py
Comment (bmorris3) [reply jdaviz/configs/default/plugins/model_fitting/fitting_backend.py:265]: I suspect that no Cubeviz behavior when trying to fit a fully-masked spectral cube will prevent users from being disappointed.

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (bmorris3) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py:848]: I agree that addition isn't right here. I think what we actually want is 
```python
spectrum.mask |= subset_mask
```
If either the spectral or the spatial coordinate has `mask==True`, we want to exclude that elements from the fit. We don't want `*` or `&` because that would only exclude cube elements where the mask is `True` for both the spectral and spatial masks.

Comment (rosteen) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py]: I believe the masks are actually integers, not booleans, hence why addition works here. Anything other than 0 is masked.

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (kecnry) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py]: I think I explained this poorly.  I'm not sure moving the entire `_apply_subset_masks` will work (cleanly) because of the way model fitting is caching the spectrum object manually outside of the components.  I was thinking more of just this single line (and the lines before it to retrieve the applicable viewer).

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (bmorris3) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py]: Just checked: the output of `get_subset_object` is an `ndarray` with `dtype('bool')`.

Comment (rosteen) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py]: I see, not what I expected. Checking the specutils docs it does allow either boolean or integer `mask` though, I remembered incorrectly that they're always integer.

Comment (kecnry) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py]: For booleans, there probably isn't any difference between `|=` and `+=`, but I think it's more readable to change (or add a comment reminding that they're booleans).

File: jdaviz/configs/default/plugins/model_fitting/fitting_backend.py
Comment (pllim) [reply jdaviz/configs/default/plugins/model_fitting/fitting_backend.py:265]: Not sure if I understand. Now that we support mask, new use cases should be tested. What if mask is very small, completely out of bounds, partial coverage, etc.

Comment (bmorris3) [reply jdaviz/configs/default/plugins/model_fitting/fitting_backend.py:265]: jdaviz does the fit via the `specutils.fitting.fit_lines` function, which [uses `astropy.fitting.LevMarLSQFitter`](https://github.com/astropy/astropy/blob/b2f33b392c9d470d7f839e6135485aac6a1c5f4a/astropy/modeling/fitting.py#L1268) under the hood, which in turn [relies on `scipy.optimize.leastsq`](https://github.com/astropy/astropy/blob/b2f33b392c9d470d7f839e6135485aac6a1c5f4a/astropy/modeling/fitting.py#L1313-L1320). The [`leastsq` docs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.leastsq.html) don't mention masking, and the [scipy source](https://github.com/scipy/scipy/blob/v1.9.3/scipy/optimize/_minpack_py.py#L279-L479) isn't illuminating either. To summarize: I don't know where the masking gets implemented.

Comment (rosteen) [reply jdaviz/configs/default/plugins/model_fitting/fitting_backend.py:265]: I think @pllim's point is most directed at spatial subsets - if the user is fitting a 10x10 square spatial subset, for example, can we speed this up but bailing early on the other 5000 spaxels rather than passing fully masked Spectrum1D objects to the lower level fitting code at all? It's the case of the whole spaxel being masked, not the whole cube being masked.

Comment (bmorris3) [reply jdaviz/configs/default/plugins/model_fitting/fitting_backend.py:265]: > It's the case of the whole spaxel being masked, not the whole cube being masked.

Ah _this_ is what I didn't understand correctly in my first reads of @pllim's comment, sorry for my misinterpretation.

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.py:831]: can this be simplified further (not sure I got all the indentation right, its hard to edit here)

```suggestion
        for subset_component in [self.spatial_subset, self.spectral_subset]:
            if subset_component is None:
                # spatial_subset only exists for cubeviz
                continue
           # only look for a mask if there is a selected subset:
           if subset_component.selected == subset_component.default_text:
               continue
            
           subset_mask = subset_component.selected_subset_mask
```

Comment (bmorris3) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py]: I had to tweak the suggestion a bunch of times so I made a separate commit: 
255d5c5.

File: jdaviz/configs/default/plugins/model_fitting/fitting_backend.py
Comment (rosteen) [inline jdaviz/configs/default/plugins/model_fitting/fitting_backend.py:382]:
```suggestion
    spx = [[(x, y) for x in range(n_x) if np.any(~mask[x, y])]
           for y in range(n_y) if np.any(~mask[:, y])]
```

I _think_ that the way you currently have it gives the opposite of what we want - `np.any` will return if any pixels in the spaxel `are` masked (since True=masked), so a fully-masked spaxel would be returned here while a fully non-masked spaxel would be not.

File: jdaviz/configs/default/plugins/model_fitting/fitting_backend.py
Comment (rosteen) [inline jdaviz/configs/default/plugins/model_fitting/fitting_backend.py:384]: This is old code but I happened to look at it and now my brain hurts 😵

File: jdaviz/configs/default/plugins/model_fitting/fitting_backend.py
Comment (bmorris3) [reply jdaviz/configs/default/plugins/model_fitting/fitting_backend.py]: Yikes, thanks for this catch. I have a bad habit of making `mask`s in my work (`True`==good) that have the opposite meaning of the numpy convention (`True`==bad). I should stop it but it's a weird habit to break.

Comment (bmorris3) [reply jdaviz/configs/default/plugins/model_fitting/fitting_backend.py]: For reasons I'll never know, it made more sense to novice-numpy-user me to write `plt.plot(x[mask], y[mask])` without needing a logical not. 🤷🏻‍♂️

**Comment by rosteen** [comment]
I ran into an error when testing. Triggered by trying to do a fit with a linear component (probably triggers with any fit) to the collapsed spectrum with a spectral subset:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File ~/opt/anaconda3/envs/viz_dev/lib/python3.10/site-packages/ipyvue/VueTemplateWidget.py:60, in Events._handle_event(self, _, content, buffers)
     58     getattr(self, "vue_" + event)(data, buffers)
     59 else:
---> 60     getattr(self, "vue_" + event)(data)

File ~/projects/jdaviz/jdaviz/configs/default/plugins/model_fitting/model_fitting.py:653, in ModelFitting.vue_apply(self, event)
    652 def vue_apply(self, event):
--> 653     self.calculate_fit()

File ~/projects/jdaviz/jdaviz/configs/default/plugins/model_fitting/model_fitting.py:650, in ModelFitting.calculate_fit(self, add_data)
    648     return self._fit_model_to_cube(add_data=add_data)
    649 else:
--> 650     return self._fit_model_to_spectrum(add_data=add_data)

File ~/projects/jdaviz/jdaviz/configs/default/plugins/model_fitting/model_fitting.py:666, in ModelFitting._fit_model_to_spectrum(self, add_data)
    663 models_to_fit = self._reinitialize_with_fixed()
    665 # Apply masks from selected spectral, spatial subsets
--> 666 self._apply_subset_masks(self._spectrum1d)
    668 try:
    669     fitted_model, fitted_spectrum = fit_model_to_spectrum(
    670         self._spectrum1d,
    671         models_to_fit,
   (...)
    674         window=self._window
    675     )

File ~/projects/jdaviz/jdaviz/configs/default/plugins/model_fitting/model_fitting.py:823, in ModelFitting._apply_subset_masks(self, spectrum)
    819 subset_mask = subset_component.selected_subset_mask
    821 if subset_mask.ndim == 1:
    822     # broadcast to the dimensions of the full cube:
--> 823     subset_mask = np.broadcast_to(
    824         subset_mask[None, None, :], spectrum.shape
    825     )
    826 elif subset_mask.shape != spectrum.flux.shape:
    827     subset_mask = np.swapaxes(subset_mask, 1, 0)

File <__array_function__ internals>:180, in broadcast_to(*args, **kwargs)

File ~/opt/anaconda3/envs/viz_dev/lib/python3.10/site-packages/numpy/lib/stride_tricks.py:413, in broadcast_to(array, shape, subok)
    367 @array_function_dispatch(_broadcast_to_dispatcher, module='numpy')
    368 def broadcast_to(array, shape, subok=False):
    369     """Broadcast an array to a new shape.
    370 
    371     Parameters
   (...)
    411            [1, 2, 3]])
    412     """
--> 413     return _broadcast_to(array, shape, subok=subok, readonly=True)

File ~/opt/anaconda3/envs/viz_dev/lib/python3.10/site-packages/numpy/lib/stride_tricks.py:349, in _broadcast_to(array, shape, subok, readonly)
    346     raise ValueError('all elements of broadcast shape must be non-'
    347                      'negative')
    348 extras = []
--> 349 it = np.nditer(
    350     (array,), flags=['multi_index', 'refs_ok', 'zerosize_ok'] + extras,
    351     op_flags=['readonly'], itershape=shape, order='C')
    352 with it:
    353     # never really has writebackifcopy semantics
    354     broadcast = it.itviews[0]

ValueError: input operand has more dimensions than allowed by the axis remapping
```

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (rosteen) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.py:821]:
```suggestion
            if subset_mask.ndim == 1 and spectrum.flux.ndim == 3:
```
This fixes the error I encountered.

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py
Comment (kecnry) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.py:737]: doesn't the cube fitting case still need to apply both spatial and spectral (or is spectral handled elsewhere here)?

Comment (kecnry) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py:737]: I also see 

```
# Reset the data mask in case we use a different subset next time
self._spectrum1d.mask = self._original_mask
```

at the end of the spectral case... I'm guessing we also need that at the end of the cube case to avoid issues when fitting with cube enabled, disabling cube fitting, and attempting to fit the same subset combos to the spectrum.

Comment (bmorris3) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py]: RE first comment: yes, this is needed! Coming in another commit.


RE second comment: In the cube fitting case, the mask gets applied to `spec`:
https://github.com/spacetelescope/jdaviz/blob/41c8235377060cb2b61c06947c68f1cc2a4f7e84/jdaviz/configs/default/plugins/model_fitting/model_fitting.py#L725
which never gets returned to the user and by construction doesn't have a mask to start. So I don't think we need to restore an original mask in that case.

Comment (kecnry) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py]: ah, I see, spectrum fitting acts on the cached `self._spectrum1d` and cube fitting acts on `spec` which is retrieved on the fly and not cached anywhere (something we might want to refactor to be less confusing eventually, but not necessary in the scope here).

**Review by pllim — COMMENTED** [review]
I feel like it is unnecessary to create a `np.zeros` mask if it is not needed. Can't we just use the "old way" if masking is unnecessary?

Anyway, I won't be here for the final review, so if two other devs are comfortable with approving, go ahead and merge when you get those two approvals. Thanks!

File: jdaviz/configs/default/plugins/model_fitting/fitting_backend.py
Comment (rosteen) [inline jdaviz/configs/default/plugins/model_fitting/fitting_backend.py:372]: Is this redundant with line 263 in `model_fitting.py`?

**Comment by bmorris3** [comment]
After discussing a remaining bug with @rosteen and possible testing improvements, I'm going to change the milestone for this PR to 3.2, so I can get a bugfix release out now, before the holiday weekend begins (shortly!). 🦃

**Comment by bmorris3** [comment]
@rosteen showed me that a simple test-case was failing, which exposed that there was a troubling bug within the PR. The problem arose because the Glue subset mask arrays are `True` for _unmasked_ values; but numpy/astropy/specutils use `True` for _masked_ values. I added a big new set of tests in [61f51a2](https://github.com/spacetelescope/jdaviz/pull/1834/commits/61f51a20e15ba860b5fcffedaa754418d5f17646) which make it clearer how the interface should work, and shows that indeed they do work that way. 

Below is a screenshot of a successful version of the demo that @rosteen was performing (which was failing before this comment):
image

**Comment by bmorris3** [comment]
As cross-referenced above, it looks like the remaining failure is related to a deprecated dtype in numpy which is being invoked by skimage, and needs to be fixed upstream. Hopefully https://github.com/scikit-image/scikit-image/pull/6633 will suffice. Have a good weekend everyone!

**Comment by orifox** [comment]
Ok, this worked nicely. Success. I approve.

**Comment by orifox** [comment]
Screen Shot 2022-11-29 at 12 07 26 PM

**Review by rosteen — APPROVED** [review]
LGTM now.

**Comment by rosteen** [comment]
@pllim A rebase would get rid of this numpy deprecation, right? Want me to do that and force push to Brett's branch, or just go ahead and merge?

**Comment by pllim** [comment]
It is not a "required to pass" job, so if you are confident that this patch will not fail in dev job, go ahead and merge.

**Comment by pllim** [comment]
If not, a rebase will get rid of the warning because I already merged Brett's patch.

**Pull request merged**
